subsys/bluetooth/host/conn.c

```diff
@@ -614,429 +592,460 @@ static inline uint16_t conn_mtu(struct bt_conn *conn)
-
-		/* destroy the buffer */
-		net_buf_unref(buf);
-
-		/* destroy the tx context (and any associated meta-data) */
-		if (tx) {
-			conn_tx_destroy(conn, tx);
-		}
+	/* The TX processor will call the `pull_cb` to get the buf */
+	if (!atomic_set(&conn->_conn_ready_lock, 1)) {
+		sys_slist_append(&bt_dev.le.conn_ready,
+				 &conn->_conn_ready);
+		LOG_DBG("raised");
+	} else {
+		LOG_DBG("already in list");
 	}
 
-	__ASSERT(sys_slist_is_empty(&conn->tx_pending), "Pending TX packets");
-	__ASSERT_NO_MSG(conn->pending_no_cb == 0);
-
-	bt_conn_reset_rx_state(conn);
-
-	k_work_reschedule(&conn->deferred_work, K_NO_WAIT);
+	/* Kick the TX processor */
+	bt_tx_irq_raise();
 }
 
-static void conn_destroy(struct bt_conn *conn, void *data)
+static bool cannot_send_to_controller(struct bt_conn *conn)
 {
-	if (conn->state == BT_CONN_CONNECTED ||
-	    conn->state == BT_CONN_DISCONNECTING) {
-		bt_conn_set_state(conn, BT_CONN_DISCONNECT_COMPLETE);
-	}
-
-	bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+	return k_sem_count_get(bt_conn_get_pkts(conn)) == 0;
 }
 
-void bt_conn_cleanup_all(void)
+static bool dont_have_methods(struct bt_conn *conn)
 {
-	bt_conn_foreach(BT_CONN_TYPE_ALL, conn_destroy, NULL);
+	return (conn->tx_data_pull == NULL) ||
+		(conn->get_and_clear_cb == NULL) ||
+		(conn->has_data == NULL);
 }
 
-static int conn_prepare_events(struct bt_conn *conn,
-			       struct k_poll_event *events)
+struct bt_conn *get_conn_ready(void)
 {
-	if (!atomic_get(&conn->ref)) {
-		return -ENOTCONN;
-	}
+	/* Here we only peek: we pop the conn (and insert it at the back if it
+	 * still has data) after the QoS function returns false.
+	 */
+	sys_snode_t *node  = sys_slist_peek_head(&bt_dev.le.conn_ready);
 
-	if (conn->state == BT_CONN_DISCONNECTED &&
-	    atomic_test_and_clear_bit(conn->flags, BT_CONN_CLEANUP)) {
-		conn_cleanup(conn);
-		return -ENOTCONN;
+	if (node == NULL) {
+		return NULL;
 	}
 
-	if (conn->state != BT_CONN_CONNECTED) {
-		return -ENOTCONN;
+	struct bt_conn *conn = CONTAINER_OF(node, struct bt_conn, _conn_ready);
+
+	if (cannot_send_to_controller(conn)) {
+		/* We will get scheduled again when the buffers are freed. */
+		LOG_DBG("no LL bufs for %p", conn);
+		return NULL;
 	}
 
-	LOG_DBG("Adding conn %p to poll list", conn);
+	if (dont_have_tx_context(conn)) {
+		/* We will get scheduled again when TX contexts are available. */
+		LOG_DBG("no TX contexts");
+		return NULL;
+	}
 
-	/* ISO Synchronized Receiver only builds do not transmit and hence
-	 * may not have any tx buffers allocated in a Controller.
-	 */
-	struct k_sem *conn_pkts = bt_conn_get_pkts(conn);
+	CHECKIF(dont_have_methods(conn)) {
+		LOG_DBG("conn %p (type %d) is missing mandatory methods",
+			conn, conn->type);
 
-	if (!conn_pkts) {
-		return -ENOTCONN;
+		return NULL;
 	}
 
-	bool buffers_available = k_sem_count_get(conn_pkts) > 0;
-	bool packets_waiting = !k_fifo_is_empty(&conn->tx_queue);
+	/* this also works for ISO */
+	if (bt_data_should_stop(conn)) {
```

Comment:
And BREDR? 

If it works for all connections, then the comment seems a bit out of place :D

Reply:
it was a debug comment, thanks for catching it.